extensions-core/caffeine-cache/src/main/java/io/druid/client/cache/CaffeineCache.java

```diff
@@ -0,0 +1,198 @@
+  @Override
+  public void close(String namespace)
+  {
+    if (config.isEvictOnClose()) {
+      cache.asMap().keySet().removeIf(key -> key.namespace.equals(namespace));
+    }
+  }
+
+  @Override
+  public io.druid.client.cache.CacheStats getStats()
+  {
+    final com.github.benmanes.caffeine.cache.stats.CacheStats stats = cache.stats();
+    final long size = cache
+        .policy().eviction()
+        .map(eviction -> eviction.isWeighted() ? eviction.weightedSize() : OptionalLong.empty())
+        .orElse(OptionalLong.empty()).orElse(-1);
+    return new io.druid.client.cache.CacheStats(
+        stats.hitCount(),
+        stats.missCount(),
+        cache.estimatedSize(),
+        size,
+        stats.evictionCount(),
+        0,
+        stats.loadFailureCount()
+    );
+  }
+
+  @Override
+  public boolean isLocal()
+  {
+    return true;
+  }
+
+  @Override
+  public void doMonitor(ServiceEmitter emitter)
+  {
+    final CacheStats oldStats = priorStats.get();
+    final CacheStats newStats = cache.stats();
+    final CacheStats deltaStats;
+    if (oldStats == null) {
+      deltaStats = newStats;
+    } else {
+      deltaStats = newStats.minus(oldStats);
+    }
+    final ServiceMetricEvent.Builder builder = ServiceMetricEvent.builder();
+    emitter.emit(builder.build("query/cache/caffeine/delta/requests", deltaStats.requestCount()));
+    emitter.emit(builder.build("query/cache/caffeine/total/requests", newStats.requestCount()));
+    emitter.emit(builder.build("query/cache/caffeine/delta/loadTime", deltaStats.totalLoadTime()));
+    emitter.emit(builder.build("query/cache/caffeine/total/loadTime", newStats.totalLoadTime()));
+    emitter.emit(builder.build("query/cache/caffeine/delta/evictionBytes", deltaStats.evictionWeight()));
+    emitter.emit(builder.build("query/cache/caffeine/total/evictionBytes", newStats.evictionWeight()));
+    if (!priorStats.compareAndSet(oldStats, newStats)) {
+      // ISE for stack trace
+      log.warn(
+          new IllegalStateException("Multiple monitors"),
+          "Multiple monitors on the same cache causing race conditions and unreliable stats reporting"
+      );
+    }
+  }
+
+  @VisibleForTesting
+  Cache<NamedKey, byte[]> getCache()
+  {
+    return cache;
+  }
+
+  private byte[] deserialize(byte[] bytes)
+  {
+    if (bytes == null) {
+      return null;
+    }
+    final int decompressedLen = ByteBuffer.wrap(bytes).getInt();
+    final byte[] out = new byte[decompressedLen];
+    LZ4_DECOMPRESSOR.decompress(bytes, Ints.BYTES, out, 0, out.length);
+    return out;
+  }
+
+  private byte[] serialize(byte[] value)
+  {
+    final int len = LZ4_COMPRESSOR.maxCompressedLength(value.length);
+    final byte[] out = new byte[len];
+    final int compressedSize = LZ4_COMPRESSOR.compress(value, 0, value.length, out, 0);
+    return ByteBuffer.allocate(compressedSize + Ints.BYTES)
+                     .putInt(value.length)
+                     .put(out, 0, compressedSize)
```

Comment:
I don't know the compressed size until I've already compressed it. As such, if I allocate a maxCompressedLength + Ints.BYTES then I've over-allocated and need to truncate the byte array, which is another copy (unless you know a way to prevent this).